devito/ir/equations/algorithms.py

```diff
@@ -95,18 +95,33 @@ def lower_exprs(expressions, **kwargs):
                   for f in retrieve_functions(expr)}
 
         # Handle Indexeds (from index notation)
-        for i in retrieve_indexed(expr):
+        for i in retrieve_indexed(expr, deep=True):
             f = i.function
 
             # Introduce shifting to align with the computational domain
             indices = [(a + o) for a, o in zip(i.indices, f._size_nodomain.left)]
 
+            tmp_mapper = {}
+            # Indexify indices of nested functions
+            for idx, index in enumerate(indices):
+                if index.is_Function and not index.is_Dimension and not index.is_Indexed:
+                    tmp_mapper = {index: index.indexify(lshift=True)}
+                    index = index.xreplace(tmp_mapper)
+
             # Apply substitutions, if necessary
             if dimension_map:
                 indices = [j.xreplace(dimension_map) for j in indices]
 
+            if tmp_mapper:
+                indices = [j.xreplace(tmp_mapper) for j in indices]
+                mapper.update(tmp_mapper)
+
             mapper[i] = f.indexed[indices]
 
+            if tmp_mapper:
+                expr = expr.xreplace(mapper)
```

Comment:
third xreplace. So three additional xreplaces for nested indexeds?

Reply:
One is redundant, removed. Now I have one for indices and one for expr. I guess I can fuse them with `xreplace_indices` ?